### ISSUE TITLE
Should we really use confint method for Bayesian credible intervals?

### ISSUE THREAD

**Issue by jgabry** [stan-dev/rstanarm#45]
@bgoodri I've been thinking recently that it might avoid a great deal of confusion if we create a separate method for obtaining so-called credible intervals rather than using R's standard `confint`. Perhaps an S3 generic `credint` and a `credint.stanreg` method? (Or just a `credint` function, skipping the S3 method stuff?) If we want to do this then I can make the change quite easily this afternoon.

Here are some reasons why I think it's worth doing this

1. The documentation for `stats::confint`, which we can't change, clearly says it's for confidence intervals. 
2. It would give us the opportunity to document `credint`, where we can make it clear what the differences are between frequentist/Bayesian confidence/credible intervals. We don't really do this anywhere at the moment but I think the differences in interpretation are important.
3. The cost is small. The only downside I can think of is that some users will try to call `confint` for a model not estimated by optimization, but we can either issue a warning to use `credint` (and allow `confint` to still be usable) or an error that forces the use of `credint`.
4. If we wanted we could provide credible intervals based on posterior quantiles (as we do now) but also give the option to compute HPD intervals (that would be valid assuming there isn't severe multimodality).

**Comment by bgoodri** [comment]
ok

**Comment by jgabry** [comment]
Do you think it's useful to do this using `credint` generic and `credint.stanreg` method or should I just make a `credint` function (or `ci`, or `credible_interval`, or something else)?

**Comment by bgoodri** [comment]
I think a plain function is fine that verifies the input is a stanreg.

**Comment by jgabry** [comment]
Ok

**Comment by andrewgelman** [comment]
I prefer the term “uncertainty interval” rather than “credibility interval” or “confidence interval.”  For one thing, when the interval is wider, it’s more uncertainty!  When the interval has zero width, it’s zero uncertainty! Etc.

**Comment by jgabry** [comment]
Ok, that seems reasonable. Unless @bgoodri objects, then I can use "uncertainty" instead of "credible." (If we do that, then to avoid confusion I'll also mention in the documentation that they are what people will see referred to as credible intervals elsewhere.)

**Comment by jgabry** [comment]
Now that I think about it, we should probably not use the function name `credint` if we're not calling them credible intervals. Do you have a function name you'd like to use if we call them "uncertainty" intervals?

**Comment by syclik** [comment]
Just in case you need a reference...
http://andrewgelman.com/2010/12/21/lets_say_uncert/

**Comment by andrewgelman** [comment]
I’nm not so steeped in R naming conventions.  I guess “uncertainty_interval” is too long?  I seem to recall that R uses “ci” so we could just use that to be consistent?

**Comment by bgoodri** [comment]
R uses confint() which we are agreed we want to avoid.

**Comment by jgabry** [comment]
Yeah I'd like to avoid confint() unless the model was estimated using MLE

**Comment by andrewgelman** [comment]
if you don’t like uncertainty_interval(), we could do bayesint() which gets the point across.

**Comment by bgoodri** [comment]
Maybe something like central_interval() or central_int(). It's not just any
interval containing 95% of the draws.

**Comment by jgabry** [comment]
If we restrict ourselves to just using posterior quantiles then that could be ok, but if we ofter HPD intervals (not saying we definitely should, but I have the code to do it if we want) then central interval isn't quite right.

**Comment by andrewgelman** [comment]
True, it’s not the shortest probability interval (spin) which is what Tian and I recommend in a recent paper.

**Comment by andrewgelman** [comment]
Our messages overlapped . . . I actually prefer shortest probability interval (spin) rather than HPD for the reason that if the density is multimodal we’re still returning just a simple interval.
A

**Comment by bgoodri** [comment]
The spin is fine with me, although for these models multimodality shouldn't
be an issue.

**Comment by andrewgelman** [comment]
I don’t know how easy it is to implement spin in practice.  In our paper, Tian and I have an algorithm that has good statistical properties but it’s a bit complicated, actually I don’t really understand exactly what we did!

**Comment by jgabry** [comment]
I'm also fine with doing the spin intervals, but if it's not trivial to compute them then maybe it's not worth it. I haven't seen the method described in the paper yet.

**Comment by bgoodri** [comment]
With 1000 draws, it is not going to be that intensive to brute force it
relative to estimating the model. If we sort the draws, then there is a
finite number of intervals that contain at least 95% of the points and then
it is just a which.min on the diff of the admissible intervals.

**Comment by andrewgelman** [comment]
Here’s the paper:
http://www.stat.columbia.edu/~gelman/research/published/spin.pdf
Short answer is that brute force intervals are a bit noisy; some smoothing helps but then things get complicated.  The method in our paper works well but it’s a bit too complicated for my taste.  I’ve wanted to go back and do more but haven’t done so.  Noisy intervals _will_ be a problem, I think, especially given that (a) in some settings we don’t recommend taking a huge number of draws, and (b) often people want 95% intervals (which I think is generally a mistake, but that’s another story).  So it could make sense to just implement central intervals for now?
A

**Comment by jgabry** [comment]
Ben, if `S` is the posterior sample size, `prob` in (0,1) is desired probability mass in interval, and `mat` is matrix of posterior draws, would this do it:

    w <- S * prob 
    mark1 <- seq_len(S - w)
    mark2 <- sel1 + w
    sorted <- apply(mat, 2L, sort, decreasing = FALSE)
    diffs <- sorted[mark2,, drop = FALSE] - sorted[mark1,, drop = FALSE]
    min_ids <- apply(diffs, 2L, which.min)
    lo <- hi <- cbind(min_ids, seq_len(ncol(mat)))
    hi[, 1L] <- hi[, 1L] + w
    ci <- cbind(sorted[lo], sorted[hi])

**Comment by jgabry** [comment]
Andrew, regardless of which intervals we end up going with, would you prefer setting the default to something other than 95% or should we go with 95% since it's what people will be expecting? We might not have too many opportunities to mess with the defaults because if we change them many users' R scripts won't reproduce the original output. Of course, using the default arguments in a script is not advisable for this reason, but a lot of people do it.

**Comment by jgabry** [comment]
Ben, I updated the code above to fix a typo (sel1 --> mark1) and added note about w not being an integer, but now I'm thinking that the code above is just computing an HPD for a unimodal distribution and not necessary the SPIn

**Comment by bgoodri** [comment]
What is sel1?

**Comment by andrewgelman** [comment]
I dunno, what do you think?  I could see making 50% the default, or I could see 95% as the default.  There’s also the idea of using the half-width of the 68% interval as  a quantile-based standard error.  Right now we’re producing a median-based standard error—I think it’s the width of the central 50% interval, appropriately scaled?  Or something like that?  I guess we should make all these decisons now!

**Comment by jgabry** [comment]
sel1 should be mark1

**Comment by jgabry** [comment]
Yeah we should make all of these sorts of decisions now because we want to release the first version and then shouldn't really change these defaults without a really good reason

**Comment by bob-carpenter** [comment]
> On Jan 6, 2016, at 5:21 PM, Jonah Gabry <notifications@github.com> wrote:
> 
> Yeah we should make all of these sorts of decisions now because we want to release the first version and then shouldn't really change these defaults without a really good reason

+1 to that!

- Bob

**Comment by bgoodri** [comment]
We decided the standard error thing a while ago. It is what

help(mad)

describes.

**Comment by jgabry** [comment]
On Wed, Jan 6, 2016 at 5:20 PM, Andrew Gelman <notifications@github.com>
wrote:

> Right now we’re producing a median-based standard error—I think it’s the
> width of the central 50% interval, appropriately scaled? Or something like
> that? I guess we should make all these decisons now!
>

We're using the median absolute deviation from the posterior median, scaled
by 1.5 (or 1.4 something I guess)

**Comment by bgoodri** [comment]
I guess we need bayesian_int() with an argument indicating whether to do
central, spin, HPD, etc. But that doesn't solve the question of defaults.

**Comment by bob-carpenter** [comment]
Maybe the first author, Ling Yiu, understands the paper, not
to mention how they got "SPIn" past Andrew as the name.

It was my idea to apply the bootstrap in the first place (I had
just been reading Efron and Tibshirani).  Then Matt expounded on
the idea, then it got passed off to Tian and Ling Yiu.  But then
I've never used "quadratic programming", which is just engineer
speak for optimization.

I've never seen the point of SPIn or HPD.  Yes, I read that paper.
Its arguments revolve around a pointer to Box and Tiao (which I
didn't follow), a graphical illustration that doesn't tickle my
intuitions in figure 1, and a comment that the HPD contains the mode,
whereas the central interval can cut off the mode, saying that the
high  density intervals are "strongly supported by inference".  But
aren't all 95% intervals equally well supported?

The paper also points out the elephant in the room, which is that
"...highest posterior density intervals is that they depend on 
parameterization."  After that, there's an appeal to "routine data
analysis", but I'm not even sure what that means.

Also, I think you have to be very careful when mixing medians and
means.  You can't get the usual MCMC std errors on medians, for
instance.

- Bob

**Comment by bgoodri** [comment]
On Wed, Jan 6, 2016 at 5:51 PM, Bob Carpenter <notifications@github.com>
wrote:

> Also, I think you have to be very careful when mixing medians and
> means. You can't get the usual MCMC std errors on medians, for
> instance.
>

True, although we are not showing the rstanarm user the MCMCSE by default.
It can be seen if you call the summary on the underyling stanfit object, in
which case it is based on the mean and variance.

**Comment by jgabry** [comment]
Regarding the default value for the interval width (regardless of interval type), we should definitely pick now if it's going to be 95%, 50% or something else. If we use a value other than 95% then we should mention in the documentation why we're not using the traditional value. In addition to not using it as the default we could also take the extra step of explicitly discouraging the use of 95%, although then we really need to explain why or point to a reference. 

Sounds like Andrew would probably prefer 50%. Anyone else have a strong preference? Andrew, do you have particular blog posts or papers on why you think 95% is a mistake? I'm ok with 50% or 95%, I just want to make sure that whatever we decide we document clearly for the user.

**Comment by bgoodri** [comment]
I'm fine with 50%

**Comment by bob-carpenter** [comment]
I don't think 95% is a "mistake" so much as it's unstable
because each end only relies on 2.5% of the draws.  

And it makes it hard to do things like check coverage compared
to a 50% interval. 

- Bob

**Comment by bgoodri** [comment]
I think we can just import SPIn::SPIn and use that.

**Comment by andrewgelman** [comment]
1.  I don’t think we’d ever be doing HPD, I think we’d only be doing central or spin.  (I’m using “spin” not as a shorthand for the particular method in paper with Tian, but rather to stand for any shortest probability interval.  The difference is that in a bimodal distribution, HPD can be two intervals, while spin will just be one.

2.  But for now I’d favor just implementing central_int to keep thing simple.

A

**Comment by andrewgelman** [comment]
I never wanted to call it SPin, that got in by accident, then there was a problem where Cran doesn’t let you change the name of a package.  If we put it into rstan we can call it whatever we want.

My original motivation for spin or hpd is the scale parameter in the 8-schools problem.  I don’t think the central interval makes sense at all for that example.

**Comment by andrewgelman** [comment]
Yes but I really don’t feel I trust that algorithm enough to use in Stan, also it’s not clear we’re doing people any favors by adding one more option, that’s why I was thinking that maybe we’re better off sticking with the central interval for now?

**Comment by jgabry** [comment]
Just so we're 100% on the same page, for the "central interval" we're just using posterior quantiles, right? Right now I have it so that given a user-specified `prob` argument, we return the interval

    a <- (1 - prob) / 2
    quantile(theta, c(a, 1 - a))

**Comment by bgoodri** [comment]
Well, since the SPIn package exists, there is basically no effort needed to
call it from bayesian_int(). Or are you saying you would prefer
bayesian_int() to output a central interval rather than a spin?

**Comment by jgabry** [comment]
If both are useful we could allow them both (defaulting to central) by providing a "type" argument for specifying the type of interval to compute.

Regarding the name, I guess we should vote. We could go with an "uncertainty"-based name, e.g.,

    uncertainty_interval()
    uncertainty_int()
    uint()

or a Bayes-based name, e.g., 

    bayesian_interval()
    bayesian_int() 
    bayes_interval()
    bayes_int() 
    bayesint()

or we could just call it `intervals()`. If we call it `central_interval()` or something like that then we're restricting it to always only offer just a central interval.

**Comment by bgoodri** [comment]
I would say use a general function name and a type argument, but I don't
have a strong preference about the function name. Maybe some variation on
prob_int()?

**Comment by jgabry** [comment]
I just pushed a prob_int function that defaults to central but had a spin
option. But if there any doubts about SPIn the safest thing to do would be
to hold off on enabling that option when we release. We can always add an
option but removing one is suboptimal.

Also I don't mind the name prob_int, but it's a bit weird if we're going to
call them uncertainty intervals rather than probabllity intervals.

**Comment by andrewgelman** [comment]
I don’t see prob_int as a name.  Why not uncertainty_int or simply bayes_int?

Also, yeah, I don’t really trust Spin so much so I’m loath to include it yet.

**Comment by jgabry** [comment]
I guess bayes_int is ok (although int makes me thing of integration rather
than interval for whatever reason). I don't love any of these names but
there's no better option then it's not a huge deal as long as we're ok with
not changing it. Ben do you have a preference between those names or
another suggestion (I guess besides prob_int).

So I guess then we'll have a type argument with only central as a
possibility for now? That's a bit strange but it does leave us the option
of adding other possibilities at some future date.

**Comment by andrewgelman** [comment]
We could do bayes_interval.

**Comment by bob-carpenter** [comment]
> On Jan 6, 2016, at 7:35 PM, Jonah Gabry <notifications@github.com> wrote:
> 
> If both are useful we could allow them both (defaulting to central) by providing a "type" argument for specifying the type of interval to compute.
> 
> Regarding the name, I guess we should vote.

Don't ever take a vote.  Solicit advice and make your best call.

- Bob

**Comment by bob-carpenter** [comment]
Any dependency is bad from first principles, but some are
worth it. 

Is there any way that SPIn breaking or changing behavior would
break whatever it is you're building?

- Bob

**Comment by andrewgelman** [comment]
I think for now we’ve settled on not linking to Spin or including this as an option.  I just don’t understand it or trust it enough.  If I did understand/trust it, I’d say we should recode it from scratch, not link to the existing package.

**Comment by bgoodri** [comment]
On Wed, Jan 6, 2016 at 11:04 PM, Bob Carpenter <notifications@github.com>
wrote:

> Any dependency is bad from first principles, but some are
> worth it.
>
> Is there any way that SPIn breaking or changing behavior would
> break whatever it is you're building?
>

Not if we have a unit test for it, in which case they can't upload a SPIn
to CRAN the breaks stuff.

How did we go from Andrew advocating SPIn to advocating against SPIn?

**Comment by bob-carpenter** [comment]
> On Jan 6, 2016, at 11:08 PM, bgoodri <notifications@github.com> wrote:
> 
> On Wed, Jan 6, 2016 at 11:04 PM, Bob Carpenter <notifications@github.com>
> wrote:
> 
> > Any dependency is bad from first principles, but some are
> > worth it.
> >
> > Is there any way that SPIn breaking or changing behavior would
> > break whatever it is you're building?
> >
> 
> Not if we have a unit test for it, in which case they can't upload a SPIn
> to CRAN the breaks stuff.

But then how do packages ever update asynchronously?  For instance,
updating ggplot as they just did must've broken umpteen gazillion
packages.  Were they just not testing?

> How did we go from Andrew advocating SPIn to advocating against SPIn?

Liking the concept in principle, but not understanding the particulars
or implementation of the SPIn package.

My own thinking is that anything that requires a long footnote to explain
it to users costs extra!

- Bob

**Comment by andrewgelman** [comment]
Andrew is capricious.  Seriously, I never advocated using the existing “SPin” code.  What I said was that, for reasons such as 8 schools, I like the idea of shortest posterior intervals as an option.  But based on our experiments and people’s known inclinations to look at the endpoints of their 95% intervals, I wouldn’t recommend the noisy raw empirical shortest posterior intervals.  I’d recommend some smoothed version.  My paper with Tian suggests that a well-behaved smoothed procedure exists, but I’d only do it for Stan if I had something cleaner than what’s in SPin.

To put it another way, my positive feeling about shortest posterior intervals is outweighed by my negative feeling about having a non-transparent procedure that constructs those intervals.  Until we have a transparent, clean, and stable shortest posterior interval calculation, I’d prefer not to offer the option at all.

Does this make sense?

A

**Comment by bgoodri** [comment]
On Wed, Jan 6, 2016 at 11:12 PM, Bob Carpenter <notifications@github.com>
wrote:

> But then how do packages ever update asynchronously? For instance,
> updating ggplot as they just did must've broken umpteen gazillion
> packages. Were they just not testing?
>

Hadley emailed all the package maintainers that depended on ggplot2 about
three months in advance and told them how to make their packages work with
ggplot2 1.x and the forthcoming ggplot2 2.0. And even then the release got
delayed twice and did break some untested stuff like what we had in rstan.

We'll have to do the same thing with rstan 3 but on a smaller scale.

**Comment by bgoodri** [comment]
On Wed, Jan 6, 2016 at 11:12 PM, Andrew Gelman <notifications@github.com>
wrote:

> To put it another way, my positive feeling about shortest posterior
> intervals is outweighed by my negative feeling about having a
> non-transparent procedure that constructs those intervals. Until we have a
> transparent, clean, and stable shortest posterior interval calculation, I’d
> prefer not to offer the option at all.
>
> Does this make sense?
>

OK.

**Comment by jgabry** [comment]
Ok so how about `bayes_interval()`, with arguments `prob` (for specifying the desired probability mass to include in the interval), `type` (which will currently only be allowed to be 'central'), and `pars` (for selecting a subset of parameters)?

Andrew, is stability the reason you prefer 50% to 95%? If there are additional reasons I can (probably should) include those in the documentation too.

**Comment by andrewgelman** [comment]
Hi, your proposal sounds good. 

Also, I prefer 50% to 95% for 3 reasons:
1.  Computational stability
2.  More intuitive evaluation (half the 50% intervals should contain the true value)
3.  A sense that in aplications it’s best to get a sense of where the parameters and predicted values will be, not to attempt an unrealistic near-certainty. 
Item 3 is kinda complicated so in documentation we could just mention 1 and 2.

**Comment by bob-carpenter** [comment]
Neat, (1) and (2) were exactly the points I was making.  I used to
try to validate with 95% intervals, but it was too hard because
there weren't enough cases that got excluded and you never knew what
to do if 4 cases out of 30 were outside the 95% intervals.

(3) is a two-edged sword because I think people will be inclined
to "read" the 50% intervals as 95% intervals out of habit, expecting
higher coverage than they have.  But I like the point about
not trying to convey an unrealistic near-certainty (which is exactly
how I think people look at 95% intervals because the p value convention
at .05).

- Bob

**Comment by betanalpha** [comment]
bayesian_int -> post_int or posterior_int

I understand the motivation for 50 vs 95, but I think having a short interval and a larger interval together is helpful (and easy to visualize). I've found that 90% intervals are usually well-behaved in practice.

**Comment by jgabry** [comment]
posterior_int or posterior_interval fits well with other function names we
have like posterior_predict. Ben, do you have a preference?

**Comment by bgoodri** [comment]
I don't mind the longer posterior _ interval since you can do tab
completion

**Comment by jgabry** [comment]
Ok let's do posterior_interval. I'll make the change.

**Comment by jgabry** [comment]
As for how we refer to the intervals in the documentation text, how about "posterior uncertainty intervals"? Also should the 'prob' argument accept a vector and return multiple intervals (kind of like rstan::summary)?

**Comment by avehtari** [comment]
I like 90% intervals
 1) more stable than 95%
 2) related to type-S errors, 95% of the mass in central interval is above the lower value (and 95% mass is below the upper value), so it is easy to see if the probability that theta>0 (or theta<0) is larger or smaller than 95%

I used to experiment with shortest intervals, but I don't use shortest intervals anymore, because in cases where they differ significantly from the central interval, any single interval is not descriptive enough and I prefer then showing several intervals or whole distribution or computing relevant probabilities, like p(theta>0).

I re-read the Spin paper, I understand the method, it's reasonable and it could be used, but I think it will be more clear to use central intervals by default and only advertise the possibility to compute spins.

**Comment by bgoodri** [comment]
Users can call it multiple times if they want multiple things. I'm starting
to be persuaded by Hadley that having functions that return different types
of objects in different circumstances is bad.

**Comment by jgabry** [comment]
Ok, I think I agree. I'll keep the prob argument as just accepting a scalar in (0,1)

**Comment by jgabry** [comment]
I think I also like Aki's idea of defaulting to 90% rather than 50% or 95% intervals. That said, I do worry that if we don't default to 95% people will just assume they're getting 95% intervals. Are we of the "it's their problem for not reading the documentation" mindset? Or should we go with 95% to help people avoid mistakes?

**Comment by bgoodri** [comment]
Maybe paste the prob into the column names.

**Comment by jgabry** [comment]
Ok, do you prefer column names like "lower90" and "upper90" or "5%" and "95%"?

**Comment by bgoodri** [comment]
5% and 95%

**Comment by jgabry** [comment]
Ok. Aki can I copy some of your comments about 90% intervals for the documentation?

**Comment by avehtari** [comment]
Sure!

**Comment by bob-carpenter** [comment]
> On Jan 7, 2016, at 9:23 AM, bgoodri <notifications@github.com> wrote:
> 
> Users can call it multiple times if they want multiple things. I'm starting
> to be persuaded by Hadley that having functions that return different types
> of objects in different circumstances is bad.

This is just basic programming.   I've been ranting about this broken
feature of R (and other scripting languages) for years.  It's why I 
never wanted a single stan() call that did both optimization and
sampling.  It's also why I very much dislike the way a fat object
can be used to return a single type of thing, but one that behaves
differently based on circumstances.  In RStan, that happens with the
fit object, because it's trying to pack in both the compiled model
and the fitted result and probably other stuff.

- Bob
